File: packages/memory-agent/src/consolidate.ts
import type { ModelProvider } from "@opentalos/core-types";
import {
  deleteMemories,
  deleteRawMemories,
  listMemoriesForTenant,
  listRawMemoriesForTenant,
  upsertMemories,
  type MemoryType,
} from "@opentalos/postgres-memory";
import type { Pool } from "pg";
import { completeText } from "./complete-text.js";

const CONSOLIDATION_SYSTEM_PROMPT = `你是一个记忆整合助手。你会看到某个用户"已确认的记忆"列表和一批"待处理的原始观察"，任务是把原始观察合并进已确认的记忆里，输出一组具体操作。

规则：
- 一条原始观察如果是全新信息，输出一个 add 操作。
- 一条原始观察如果是对已有记忆的补充/修正/推翻，输出一个 update（改写这条记忆）或 delete（这条记忆已经不成立了）操作，引用已有记忆的 id。
- 一条原始观察如果跟已有记忆完全重复，不用输出任何操作（忽略即可）。
- profile/preference 类型的记忆一般是稳定的，不要轻易 delete；context 类型的记忆容易过时，看起来不再成立就应该 delete。

只输出一个 JSON 对象，不要有任何其他文字，格式：
{"add": [{"type": "profile"|"preference"|"context", "title": "...", "content": "..."}], "update": [{"id": "...", "content": "..."}], "delete": ["id1", "id2"]}
如果没有任何要做的操作，输出 {"add": [], "update": [], "delete": []}`;

interface ConsolidationResult {
  add: { type: MemoryType; title: string; content: string }[];
  update: { id: string; content: string }[];
  delete: string[];
}

const VALID_MEMORY_TYPES = new Set<string>(["profile", "preference", "context"]);

function isValidNewMemory(item: unknown): item is ConsolidationResult["add"][number] {
  return (
    !!item &&
    typeof item === "object" &&
    VALID_MEMORY_TYPES.has((item as Record<string, unknown>).type as string) &&
    typeof (item as Record<string, unknown>).title === "string" &&
    typeof (item as Record<string, unknown>).content === "string"
  );
}

function isValidMemoryUpdate(item: unknown): item is ConsolidationResult["update"][number] {
  return (
    !!item &&
    typeof item === "object" &&
    typeof (item as Record<string, unknown>).id === "string" &&
    typeof (item as Record<string, unknown>).content === "string"
  );
}

/** 从"已知是数组"的原始数组里过滤出形状合法的条目，把非法条目记录下来再丢弃——而不是直接
 * cast 放行。 memories.type 在 schema 里就是普通 TEXT 字段，数据库层面不会拒绝一个模型幻觉出来的
 * type 值，所以这一步是唯一的把关点。 */
function filterValid<T>(items: unknown[], isValid: (item: unknown) => item is T, label: string): T[] {
  const valid: T[] = [];
  for (const item of items) {
    if (isValid(item)) {
      valid.push(item);
    } else {
      console.error(`memory-agent: consolidation model returned an invalid ${label} item, dropping it: ${JSON.stringify(item).slice(0, 200)}`);
    }
  }
  return valid;
}

/** 任何解析失败都当成"这次什么都不做"，而不是抛出异常——原始观察本身不会丢（还留在
 * raw_memories 里，见下面的写回逻辑：只有真正跑完一次整合，处理过的原始观察才会被清空），下一
 * 次整合周期会重新尝试。跟 extraction.ts 的 parseExtractionResult 一样，区分"模型合法地说无操作"
 * 和"模型输出解析失败/形状不对"两种情况并分别记录日志，方便排查持续性的 prompt 问题。 */
function parseConsolidationResult(raw: string): ConsolidationResult {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    console.error(`memory-agent: consolidation model returned non-JSON output: ${raw.slice(0, 200)}`);
    return { add: [], update: [], delete: [] };
  }
  if (!parsed || typeof parsed !== "object") {
    console.error(`memory-agent: consolidation model returned a non-object JSON value: ${raw.slice(0, 200)}`);
    return { add: [], update: [], delete: [] };
  }
  const obj = parsed as Record<string, unknown>;
  if (!Array.isArray(obj.add) || !Array.isArray(obj.update) || !Array.isArray(obj.delete)) {
    console.error(`memory-agent: consolidation model returned an unexpected shape (missing/non-array add/update/delete): ${raw.slice(0, 200)}`);
  }
  const add = Array.isArray(obj.add) ? filterValid(obj.add, isValidNewMemory, "add") : [];
  const update = Array.isArray(obj.update) ? filterValid(obj.update, isValidMemoryUpdate, "update") : [];
  const del = Array.isArray(obj.delete) ? filterValid(obj.delete, (id): id is string => typeof id === "string", "delete") : [];
  return { add, update, delete: del };
}

export async function consolidateMemoriesForTenant(
  pool: Pool,
  provider: ModelProvider,
  tenantId: string,
  options?: { signal?: AbortSignal },
): Promise<void> {
  const [existing, raw] = await Promise.all([listMemoriesForTenant(pool, tenantId), listRawMemoriesForTenant(pool, tenantId)]);
  if (raw.length === 0) return;

  // 一旦某条 delete 提交，被删的记忆就不会再出现在下一轮传给模型的"已确认的记忆"里——如果后续
  // 步骤（比如 deleteRawMemories）在这之后失败，触发过这条 delete 的原始观察下次重跑时会被当成
  // "全新信息"重新解读，有可能把刚删掉的内容悄悄地重新加回来。
  const userPrompt =
    `已确认的记忆：\n${existing.map((m) => `- [${m.id}] [${m.type}] ${m.title}: ${m.content}`).join("\n") || "（无）"}\n\n` +
    `待处理的原始观察：\n${raw.map((r) => `- ${r.content}`).join("\n")}`;
  const rawResult = await completeText(provider, CONSOLIDATION_SYSTEM_PROMPT, userPrompt, options);
  const result = parseConsolidationResult(rawResult);

  if (result.add.length > 0 || result.update.length > 0) {
    await upsertMemories(pool, tenantId, { newEntries: result.add, updates: result.update });
  }
  if (result.delete.length > 0) {
    await deleteMemories(pool, tenantId, result.delete);
  }
  // 注意：upsertMemories/deleteMemories/deleteRawMemories 是三个独立事务，整体并不原子——如果在
  // 它们之间崩溃，重试时对同一个 title 再次 add 会撞上 (tenantId, title) 唯一约束，且没有退避/去重
  // 机制，存在活锁风险（留给调用方，即 Task 6/7 的 worker 处理）。
  // 处理过的 raw_memories 全部清空，不管模型是否采纳——这批已经被完整看过一遍，留着没有增量价值
  // （见 spec 里"整合"一节的明确设计决定）。
  await deleteRawMemories(pool, tenantId, raw.map((r) => r.id));
}

File: services/memory-service/app/consolidate.py
import asyncio
import json
import logging
from typing import Any

from psycopg_pool import AsyncConnectionPool

from db.bulk import (
    MemoryUpdate,
    NewMemory,
    delete_memories,
    delete_raw_memories,
    list_memories_for_tenant,
    list_raw_memories_for_tenant,
    upsert_memories,
)
from providers.base import TextCompletionProvider

logger = logging.getLogger(__name__)

CONSOLIDATION_SYSTEM_PROMPT = """你是一个记忆整合助手。你会看到某个用户"已确认的记忆"列表和一批"待处理的原始观察"，任务是把原始观察合并进已确认的记忆里，输出一组具体操作。

规则：
- 一条原始观察如果是全新信息，输出一个 add 操作。
- 一条原始观察如果是对已有记忆的补充/修正/推翻，输出一个 update（改写这条记忆）或 delete（这条记忆已经不成立了）操作，引用已有记忆的 id。
- 一条原始观察如果跟已有记忆完全重复，不用输出任何操作（忽略即可）。
- profile/preference 类型的记忆一般是稳定的，不要轻易 delete；context 类型的记忆容易过时，看起来不再成立就应该 delete。

只输出一个 JSON 对象，不要有任何其他文字，格式：
{"add": [{"type": "profile"|"preference"|"context", "title": "...", "content": "..."}], "update": [{"id": "...", "content": "..."}], "delete": ["id1", "id2"]}
如果没有任何要做的操作，输出 {"add": [], "update": [], "delete": []}"""

VALID_MEMORY_TYPES = {"profile", "preference", "context"}


def _is_valid_new_memory(item: Any) -> bool:
    return (
        isinstance(item, dict)
        and item.get("type") in VALID_MEMORY_TYPES
        and isinstance(item.get("title"), str)
        and isinstance(item.get("content"), str)
    )


def _is_valid_memory_update(item: Any) -> bool:
    return isinstance(item, dict) and isinstance(item.get("id"), str) and isinstance(item.get("content"), str)


def _filter_valid(items: list[Any], is_valid: Any, label: str) -> list[Any]:
    valid = []
    for item in items:
        if is_valid(item):
            valid.append(item)
        else:
            logger.error(
                "memory-service: consolidation model returned an invalid %s item, dropping it: %s",
                label,
                json.dumps(item, ensure_ascii=False)[:200],
            )
    return valid


def _parse_consolidation_result(raw: str) -> tuple[list[NewMemory], list[MemoryUpdate], list[str]]:
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        logger.error("memory-service: consolidation model returned non-JSON output: %s", raw[:200])
        return [], [], []
    if not isinstance(parsed, dict):
        logger.error("memory-service: consolidation model returned a non-object JSON value: %s", raw[:200])
        return [], [], []

    add_raw = parsed.get("add") if isinstance(parsed.get("add"), list) else []
    update_raw = parsed.get("update") if isinstance(parsed.get("update"), list) else []
    delete_raw = parsed.get("delete") if isinstance(parsed.get("delete"), list) else []
    if not isinstance(parsed.get("add"), list) or not isinstance(parsed.get("update"), list) or not isinstance(parsed.get("delete"), list):
        logger.error(
            "memory-service: consolidation model returned an unexpected shape (missing/non-array add/update/delete): %s",
            raw[:200],
        )

    valid_add = _filter_valid(add_raw, _is_valid_new_memory, "add")
    valid_update = _filter_valid(update_raw, _is_valid_memory_update, "update")
    valid_delete = _filter_valid(delete_raw, lambda item: isinstance(item, str), "delete")

    return (
        [NewMemory(type=item["type"], title=item["title"], content=item["content"]) for item in valid_add],
        [MemoryUpdate(id=item["id"], content=item["content"]) for item in valid_update],
        valid_delete,
    )


async def consolidate_memories_for_tenant(
    pool: AsyncConnectionPool, provider: TextCompletionProvider, tenant_id: str
) -> None:
    existing, raw = await asyncio.gather(
        list_memories_for_tenant(pool, tenant_id), list_raw_memories_for_tenant(pool, tenant_id)
    )
    if not raw:
        return

    existing_lines = "\n".join(f"- [{m.id}] [{m.type}] {m.title}: {m.content}" for m in existing) or "（无）"
    raw_lines = "\n".join(f"- {r.content}" for r in raw)
    user_prompt = f"已确认的记忆：\n{existing_lines}\n\n待处理的原始观察：\n{raw_lines}"

    raw_result = await provider.complete_text(CONSOLIDATION_SYSTEM_PROMPT, user_prompt)
    add, update, delete = _parse_consolidation_result(raw_result)

    if add or update:
        await upsert_memories(pool, tenant_id, add, update)
    if delete:
        await delete_memories(pool, tenant_id, delete)

    # 处理过的 raw_memories 全部清空，不管模型是否采纳——这批已经被完整看过一遍，留着没有增量
    # 价值。注意 upsert_memories/delete_memories/delete_raw_memories 是三个独立事务，整体不原子——
    # 如果在它们之间崩溃，重试时对同一个 title 再次 add 会撞上 (tenant_id, title) 唯一约束。
    await delete_raw_memories(pool, tenant_id, [r.id for r in raw])
